---
title: "2022_08_08_rChIPcandidate_examination"
output: html_document
editor_options: 
  chunk_output_type: console
---


---
title: "00_new_candidate_qc"
author: "JR"
date: "7/26/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = F)
library(GenomicRanges)
library(tidyverse)
# library(Gviz)
library(IRanges)
source("/scratch/Shares/rinn/tenaya/util/intersect_functions.R")
source("/scratch/Shares/rinn/tenaya/util/plotting_functions.R")
library(valr)
```

First let's look at CTCF to compare the "good" vs "bad" chip

# CTCF (10-20, 20-60 shearing) = "good"

```{r CTCF}
# TODO More analyses per candidate :
# (1) Feature counts of chip peaks for RNA in frip. 
# 
# Check in RDeep for shifting of candidates
# 
# Width vs quality: did this and pretty flat across widths (so width doesn't influence peak sig as expected by MACS)
# 
# First we need promoters let's make them quick: (actually takes 5 min or so)
gencode_gr <- rtracklayer::import("/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/gencode.v32.annotation.gtf")
# create promoters 
promoters <- promoters(gencode_gr[gencode_gr$type == "gene"], upstream = 1000, downstream = 1000) 

# CTCF good vs bad signal/quality vs overlap (not sure super relevant form prelim look)
ctcf_peaks <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_10-20_shearing_R1_peaks.broadPeak")
ctcf_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
ctcf_metaplot_df <- profile_tss(ctcf_peaks , promoters)
# plotting
ggplot(ctcf_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
# Nice
length(ctcf_peaks)
# CTCF 20-60 shearing
ctcf_20_60 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_20-60_shearing_R1_peaks.broadPeak")
# 
ctcf20_60_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
ctcf20_60_metaplot_df <- profile_tss(ctcf_20_60 , promoters)
# plotting
ggplot(ctcf20_60_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
# TODO make consensus peaks and replot.
```

```{r}
ctcf_peaks <- read_table("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_10-20_shearing_R1_peaks.broadPeak", col_names = F) %>%
  mutate(width = X3 - X2)
ggplot(ctcf_peaks, aes(x = log10(width), y = X9)) +
  geom_point() +
  ylim(0,200)
ctcf_peaks_bad <- read_table("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_no_rnase_R1_peaks.broadPeak", col_names = F) %>%
  mutate(width = X3 - X2)
ggplot(ctcf_peaks_bad, aes(x = log10(width), y = X9)) +
  geom_point() +
  ylim(0,200)
```

# comparing a good CTCF to bad one

```{r CTCF good vs bad}
ctcf_peaks <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_10-20_shearing_R1_peaks.broadPeak")
ctcf_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
ctcf_metaplot_df <- profile_tss(ctcf_peaks , promoters)
# plotting
ggplot(ctcf_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
# bad
ctcf_bad_peaks <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/CTCF_no_rnase_R1_peaks.broadPeak")
ctcf_bad_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
ctcf_bad_metaplot_df <- profile_tss(ctcf_bad_peaks , promoters)
# plotting
ggplot(ctcf_bad_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
length(ctcf_bad_peaks)
# compare overlaps with old ctcf and new one
# plus rnase metaplot and overlaps and good one 
# plotting them together
ctcf_bad_metaplot_df$quality <- "bad"
ctcf_metaplot_df$quality <- "good"
ctcf_combined_metaplot_profile <- bind_rows(ctcf_bad_metaplot_df, ctcf_metaplot_df)
ggplot(ctcf_combined_metaplot_profile, 
       aes(x = x, y = dens, color = quality)) +
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle("CTCF Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-3000, 0, 3000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency") + 
  scale_color_manual(values = c("#424242","#a8404c"))
# checking num of peaks and overlaps
length(ctcf_bad_peaks)
# good peaks
length(ctcf_peaks)
# length of overlap
ctcf_good_bad_ov <- subsetByOverlaps(ctcf_bad_peaks, ctcf_peaks)
# % overlap
length(ctcf_good_bad_ov)/ length(ctcf_bad_peaks)
```


# RBP1
Trying RBP1 as comparison
```{r}
# loading in peak file
RBP1_peaks <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/RBP1_no_rnase_R1_peaks.broadPeak")
RBP1_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
# running profile TSS 
RBP1_metaplot_df <- profile_tss(RBP1_peaks , promoters)
# plotting metaplot
ggplot(RBP1_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
```




SAF-A (130 peaks)

```{r}
# loading in SAFA rep1
safa_peaks <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/SAFA_no_rnase_R1_peaks.broadPeak")
# metaplot_df
safa_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
safa_metaplot_df <- profile_tss(safa_peaks , promoters)
# plotting
ggplot(safa_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
# wow that looks really bad
length(safa_peaks)
# trying the other replicate
safa_rnase_peaks <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/SAFA_RNAse_R1_peaks.broadPeak")
safa_rnase_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
safa_rnase_metaplot_df <- profile_tss(safa_peaks , promoters)
# plotting
ggplot(safa_rnase_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
# wow that looks really bad
length(safa_rnase_peaks)
```



# Now for the new candidates: HDAC1, PHF8 and CBX3, (STAG2/1)




First going to encode portal and grabbing latest "consensus peaks" 
between replicates for each candidate.

# Candidate 1: HDAC1
First download the consensus peak files of each candidate using wget.

```{bash}
# change to working dir
cd /scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/HDAC1
# download 
HDAC1: wget https://www.encodeproject.org/files/ENCFF432KJA/@@download/ENCFF432KJA.bed.gz
# there are other we can add but look similar
# They are all here:
# https://www.encodeproject.org/search/?type=Experiment&control_type!=*&status=released&perturbed=false&target.label=HDAC1&biosample_ontology.term_name=K562&assay_title=TF+ChIP-seq&biosample_ontology.classification=cell%20line
# example:
# HDAC1_2: https://www.encodeproject.org/files/ENCFF661VOO/@@download/ENCFF661VOO.bed.gz
# let's unzip the peak file
gunzip ENCFF432KJA.bed.gz
```


# HDAC1 peaks to Granges
```{r import peak file}
#HDAC1_peaks <- read.table("/scratch/Shares/rinn/tenaya/rChIP/ENCODE_downloads/ENCFF432KJA.bed.gz")
HDAC1_peaks <- read_tsv("/scratch/Shares/rinn/tenaya/rChIP/ENCODE_downloads/ENCFF432KJA.bed.gz", col_names = F)
# col values are :
# Chromosome, start, stop, name, score, strand, signalValue, pValue, qValue, peak_center
# adding colnames 
names(HDAC1_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
# Note this is bed+4 format / .broadPeak files are .bed+3 -- no peak center
# the rest is the same
# Nice the data frame is reeadable and reachable 
# Now we want to find some good peaks in the raw data by eye
# We have tracks in UCSC or load in bigWig into IGV
# Mostly we want to find big, sig peaks that are in good areas of genome
# converting to Granges from DF
HDAC1_1_gr <- GRanges(HDAC1_peaks$chromosome, IRanges(HDAC1_peaks$start, HDAC1_peaks$end))
# test <- GRanges(seqnames = HDAC1_peaks$chromosome,
#                 ranges = IRanges(start=HDAC1_peaks$start,end=HDAC1_peaks$end))
```

# Writing the HDAC1 peaks into a csv for Primer Examination 
```{r}
write.csv(HDAC1_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//HDAC1_peaks_ENCFF432KJA.csv",row.names = TRUE)
#we want the higher qValue. so sort
#then look at the positions in the UCSC browser
#Narrow window
#Get Sequence
#Primer3 to design the primers
#Check primers with NCBI
#Check primers with IDT
```


# HDAC1 metaplot of peaks
```{r}
# First we need promoters let's make them quick: (actually takes 5 min or so)
#gencode_gr <- rtracklayer::import("/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/gencode.v32.annotation.gtf")
# create promoters 
#promoters <- promoters(gencode_gr[gencode_gr$type == "gene"], upstream = 1000, downstream = 1000)
# making empty data frame for profile_tss to populate
HDAC1_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
# running profile TSS 
HDAC1_metaplot_df <- profile_tss(HDAC1_1_gr , promoters)
ggplot(HDAC1_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
length(HDAC1_1_gr)
```

# Valr metaplot stuff for HDAC1: Peak center overlaps
```{r }
# Let's get the promoters in the valr format.
# We'll just get the TSS
promoter_bed <- promoters(gencode_gr[gencode_gr$type == "gene"], upstream = 1000, downstream = 1000) %>%
  gr_to_bed() %>%
  mutate(tss = start + ((end-start)/2))
# We're just going to take the peak center.
HDAC1_peaks  <- read_bed("/scratch/Shares/rinn/tenaya/rChIP/ENCODE_downloads/ENCFF432KJA.bed.gz") %>%
  mutate(peak_center = start + X10)
hdac_tss_ov <- bed_intersect(HDAC1_peaks, promoter_bed) %>%
  mutate(meta_position = peak_center.x - tss.y,
         meta_position = ifelse(strand.y == "-", meta_position * -1, meta_position))
hdac_tss_meta_peak_center <- hdac_tss_ov %>%
  filter(meta_position >= -1000, meta_position <= 1000) %>%
  group_by(meta_position) %>%
  summarize(npeaks = n())
ggplot(hdac_tss_meta_peak_center, aes(x = meta_position, y = npeaks)) +
  geom_point() +
  geom_smooth(span = 0.2)
```


# Candidate 2: PHF8
```{r PHF8 }
#download from ENCODE:
# wget https://www.encodeproject.org/files/ENCFF981ISM/@@download/ENCFF981ISM.bed.gz
# reading in peaks from encode
PHF8_peaks <- read_tsv("/scratch/Shares/rinn/tenaya/rChIP/ENCODE_downloads/ENCFF981ISM.bed.gz", col_names = F)
# adding colnames 
names(PHF8_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
# converting to Granges
PHF8_gr <- GRanges(PHF8_peaks$chromosome, IRanges(PHF8_peaks$start, PHF8_peaks$end))
# creating metaplot DF and running profile_tss
PHF8_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
PHF8_metaplot_df <- profile_tss(PHF8_gr, promoters)
# plotting
ggplot(PHF8_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
length(PHF8_gr)
subsetByOverlaps(PHF8_gr, promoters)
```

#Writing CSV for PHF8 rChIP Primer Design
```{r}
write.csv(PHF8_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//PHF8_peaks_ENCFF981ISM.csv",row.names = TRUE)
```


# Valr metaplot stuff for PHF8
```{r valr meta plot PHF8}
# Let's get the promoters in the valr format.
# We'll just get the TSS
# tss <- promoters(gencode_gr[gencode_gr$type == "gene"], upstream = 0, downstream = 0) %>%
#   gr_to_bed()
# We're just going to take the peak center.
PHF8_peaks  <- read_bed("/scratch/Shares/rinn/tenaya/rChIP/ENCODE_downloads/ENCFF981ISM.bed.gz") %>%
  mutate(peak_center = start + X10)
PHF8_tss_ov <- bed_intersect(PHF8_peaks, promoter_bed) %>%
  mutate(meta_position = peak_center.x - tss.y,
         meta_position = ifelse(strand.y == "-", meta_position * -1, meta_position))
PHF8_tss_meta_peak_center <- PHF8_tss_ov %>%
  filter(meta_position >= -1000, meta_position <= 1000) %>%
  group_by(meta_position) %>%
  summarize(npeaks = n())
ggplot(PHF8_tss_meta_peak_center, aes(x = meta_position, y = npeaks)) +
  geom_point() +
  geom_smooth(span = 0.2)
```



# Candidate 3 : CBX3

```{r}
# download data from ENCODE narrow peak
# wget https://www.encodeproject.org/files/ENCFF068OEJ/@@download/ENCFF068OEJ.bed.gz
# reading in peaks from encode
CBX3_peaks <- read_tsv("/scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/CBX3/ENCFF068OEJ.bed.gz", col_names = F)
# adding colnames 
names(CBX3_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
# converting to Granges
CBX3_gr <- GRanges(CBX3_peaks$chromosome, IRanges(CBX3_peaks$start, CBX3_peaks$end))
# creating metaplot DF and running profile_tss
CBX3_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
CBX3_metaplot_df <- profile_tss(CBX3_gr, promoters)
# plotting
ggplot(CBX3_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
nrow(CBX3_peaks)
# only 750 or so peaks 
subsetByOverlaps(CBX3_gr, promoters)
# trying second replicate (richard meyers)
# https://www.encodeproject.org/files/ENCFF386ZWO/@@download/ENCFF386ZWO.bed.gz
CBX3_peaks_2 <- read_tsv("/scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/CBX3/ENCFF386ZWO.bed.gz", col_names = F)
# adding colnames 
names(CBX3_peaks_2) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
# converting to Granges
CBX3_gr_2 <- GRanges(CBX3_peaks_2$chromosome, IRanges(CBX3_peaks_2$start, CBX3_peaks_2$end))
# creating metaplot DF and running profile_tss
CBX3_metaplot_df_2 <- data.frame(x = integer(), dens = numeric(), dbp = character())
CBX3_metaplot_df_2 <- profile_tss(CBX3_gr_2, promoters)
# plotting
ggplot(CBX3_metaplot_df_2, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
nrow(CBX3_peaks_2)
# 24,848
subsetByOverlaps(CBX3_gr_2, promoters)
# for comparison
ggplot(CBX3_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
```

#Writing CSV for Cbx3 rChIP Primer Design
```{r}
write.csv(CBX3_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//Cbx3_peaks_ENCFF068OEJ.csv",row.names = TRUE)
```

# Valr metaplot stuff for CBX3
```{r valr meta plot PHF8}
# Let's get the promoters in the valr format.
# We'll just get the TSS
# Let's look at the 
# tss <- promoters(gencode_gr[gencode_gr$type == "gene"], upstream = 0, downstream = 0) %>%
#   gr_to_bed()
# We're just going to take the peak center.
CBX3_peaks  <- read_bed("/scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/CBX3/ENCFF068OEJ.bed.gz") %>%
  mutate(peak_center = start + X10)
# Second replicate
CBX3_peaks_2  <- read_bed("/scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/CBX3/ENCFF386ZWO.bed.gz") %>%
  mutate(peak_center = start + X10)
# Tss overlap 
CBX3_tss_ov <- bed_intersect(CBX3_peaks, promoter_bed) %>%
  mutate(meta_position = peak_center.x - tss.y,
         meta_position = ifelse(strand.y == "-", meta_position * -1, meta_position))
# second replicate
CBX3_tss_ov_2 <- bed_intersect(CBX3_peaks_2, promoter_bed) %>%
  mutate(meta_position = peak_center.x - tss.y,
         meta_position = ifelse(strand.y == "-", meta_position * -1, meta_position)) %>%
  unite(peak_coord, chrom, start.x, end.x, remove = FALSE)
length(unique(CBX3_tss_ov_2$peak_coord))
hmm <- findOverlaps(GRanges(CBX3_peaks_2$chrom, ranges = IRanges(CBX3_peaks_2$start, CBX3_peaks_2$end)), 
                         GRanges(promoter_bed$chrom, ranges = IRanges(promoter_bed$start, promoter_bed$end)))
table(table(hmm@from))
# summarize
CBX3_tss_meta_peak_center <- CBX3_tss_ov %>%
  filter(meta_position >= -1000, meta_position <= 1000) %>%
  group_by(meta_position) %>%
  summarize(npeaks = n())
# summarize second rep
CBX3_tss_meta_peak_center_2 <- CBX3_tss_ov_2 %>%
  filter(meta_position >= -1000, meta_position <= 1000) %>%
  group_by(meta_position) %>%
  summarize(npeaks = n())
# plotting
ggplot(CBX3_tss_meta_peak_center, aes(x = meta_position, y = npeaks)) +
  geom_point() +
  geom_smooth(span = 0.2)
# plotting second rep
ggplot(CBX3_tss_meta_peak_center_2, aes(x = meta_position, y = npeaks)) +
  geom_point() +
  geom_smooth(span = 0.2)
test <- subsetByOverlaps(GRanges(CBX3_peaks_2$chrom, ranges = IRanges(CBX3_peaks_2$start, CBX3_peaks_2$end)), 
                         GRanges(promoter_bed$chrom, ranges = IRanges(promoter_bed$start, promoter_bed$end)))
length(test)
# 6,520 peaks overlap promoters
nrow(CBX3_peaks_2)
# 24,848
```

# all combined candidate metaplots

```{r}
HDAC1_metaplot_df$dbp <- "HDAC1"
PHF8_metaplot_df$dbp <- "PHF8"
CBX3_metaplot_df$dbp <- "CBX3_1"
CBX3_metaplot_df_2$dbp <- "CBX3_2"
ctcf_metaplot_df$dbp <- "CTCF"
combined_metaplot_profile <- bind_rows(HDAC1_metaplot_df, PHF8_metaplot_df, CBX3_metaplot_df, CBX3_metaplot_df_2, ctcf_metaplot_df)
ggplot(combined_metaplot_profile, 
       aes(x = x, y = dens, color = dbp)) +
  geom_vline(xintercept = 0, lty = 2) + 
  geom_line(size = 1.5) + 
  ggtitle(" Promoter Metaplot") + 
  scale_x_continuous(breaks = c(-1000, 0, 1000),
                     labels = c("-3kb", "TSS", "+3kb"),
                     name = "") + 
  ylab("Peak frequency") 
  # scale_color_manual(values = c("#424242","#a8404c"))
```


# stag1 

```{r}
# wget https://www.encodeproject.org/files/ENCFF921BXP/@@download/ENCFF921BXP.bed.gz
# reading in peaks from encode
STAG_peaks <- read_tsv("/scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/STAG1_2/ENCFF921BXP.bed.gz", col_names = F)
# adding colnames 
names(STAG_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
# converting to Granges
STAG_gr <- GRanges(STAG_peaks$chromosome, IRanges(STAG_peaks$start, STAG_peaks$end))
# creating metaplot DF and running profile_tss
STAG_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
STAG_metaplot_df <- profile_tss(STAG_gr, promoters)
# plotting
ggplot(STAG_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
nrow(STAG_peaks)
```

#Rbbp5

```{r}
# wget https://www.encodeproject.org/files/ENCFF666PCE/@@download/ENCFF666PCE.bed.gz
 
rbbp5_peaks <- read_tsv("/scratch/Shares/rinn/JR/rchip/analysis/00_NEW_candidate_qc/data/rbbp5/ENCFF666PCE.bed.gz", col_names = F)
# adding colnames 
names(rbbp5_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
# converting to Granges
rbbp5_gr <- GRanges(rbbp5_peaks$chromosome, IRanges(rbbp5_peaks$start, rbbp5_peaks$end))
# creating metaplot DF and running profile_tss
rbbp5_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
rbbp5_metaplot_df <- profile_tss(rbbp5_gr, promoters)
# plotting
ggplot(rbbp5_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
nrow(rbbp5_peaks)
# only 750 or so peaks 
subsetByOverlaps(rbbp5_gr, promoters)
```

#Writing CSV for Rbbp5 rChIP Primer Design
```{r}
write.csv(rbbp5_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//Rbbp5_peaks_ENCFF666PCE.csv",row.names = TRUE)
```







check consistency between replicates
# replicate consistency
```{r replicate overlaps}
# TODO check manually by subsetByOverlaps of each peak file 
# TODO count # peaks and overlaps
# could add promoter overlap?
```


#EZH2
```{r}
#Go to ENCODE
#Find a K562 ChIP experiment
#Go to File
#Click on the bed narrow peak
#Right click on the download file button, copy link. 

#BASH in your ENCODE downloads file
# wget https://www.encodeproject.org/files/ENCFF804RVA/@@download/ENCFF804RVA.bed.gz
ezh2_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/ENCODE_downloads/ENCFF804RVA.bed.gz", col_names = F)
# adding colnames 
names(ezh2_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
#Write into a csv
write.csv(ezh2_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//Ezh2_peaks_ENCFF804RVA.csv",row.names = TRUE)

# converting to Granges
ezh2_gr <- GRanges(ezh2_peaks$chromosome, IRanges(ezh2_peaks$start, ezh2_peaks$end))
# creating metaplot DF and running profile_tss
ezh2_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
ezh2_metaplot_df <- profile_tss(ezh2_gr, promoters)
# plotting
ggplot(ezh2_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
nrow(ezh2_peaks)
# only 750 or so peaks 
subsetByOverlaps(ezh2_gr, promoters)
```

#KDM1A
```{r}
# wget https://www.encodeproject.org/files/ENCFF829FZW/@@download/ENCFF829FZW.bed.gz
kdm1a_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/ENCODE_downloads/ENCFF829FZW.bed.gz", col_names = F)
names(kdm1a_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
write.csv(kdm1a_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//Kdm1a_peaks_ENCFF829FZW.csv",row.names = TRUE)
#31135 peaks as a heads up

# converting to Granges
kdm1a_gr <- GRanges(kdm1a_peaks$chromosome, IRanges(kdm1a_peaks$start, kdm1a_peaks$end))
# creating metaplot DF and running profile_tss
kdm1a_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
kdm1a_metaplot_df <- profile_tss(kdm1a_gr, promoters)
# plotting
ggplot(kdm1a_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
nrow(kdm1a_peaks)
# only 750 or so peaks 
subsetByOverlaps(kdm1a_gr, promoters)

```


```{r}
#EWSR1 ChIP
# wget https://www.encodeproject.org/files/ENCFF924FYI/@@download/ENCFF924FYI.bed.gz
#There is only one replicate

ewsr1_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/ENCODE_downloads/ENCFF924FYI.bed.gz", col_names = F)
names(ewsr1_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')
write.csv(ewsr1_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//ewsr1_peaks_ENCFF924FYI.csv",row.names = TRUE)
#1264 peaks as a heads up

# converting to Granges
ewsr1_gr <- GRanges(ewsr1_peaks$chromosome, IRanges(ewsr1_peaks$start, ewsr1_peaks$end))
# creating metaplot DF and running profile_tss
ewsr1_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
ewsr1_metaplot_df <- profile_tss(ewsr1_gr, promoters)
# plotting
ggplot(ewsr1_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)
nrow(ewsr1_peaks)
# only 750 or so peaks 
subsetByOverlaps(ewsr1_gr, promoters)
```


```{r}
#KHSRP
# wget https://www.encodeproject.org/files/ENCFF525XXS/@@download/ENCFF525XXS.bed.gz
khsrp_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/ENCODE_downloads/ENCFF525XXS.bed.gz", col_names = F)

names(khsrp_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')

khsrp_gr <- GRanges(khsrp_peaks$chromosome, IRanges(khsrp_peaks$start, khsrp_peaks$end))
# creating metaplot DF and running profile_tss
khsrp_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
khsrp_metaplot_df <- profile_tss(khsrp_gr, promoters)
# plotting
ggplot(khsrp_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)

write.csv(khsrp_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//khsrp_peaks_ENCFF525XXS.csv",row.names = TRUE)
```


#YYI (Suggestion from Taatjes during the retreat)
```{r}
# wget https://www.encodeproject.org/files/ENCFF328XKC/@@download/ENCFF328XKC.bed.gz
#Above is Richard Meyer's
yy1_meyer_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/ENCODE_downloads/ENCFF328XKC.bed.gz", col_names = F)

names(yy1_meyer_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')

yy1_meyer_gr <- GRanges(yy1_meyer_peaks$chromosome, IRanges(yy1_meyer_peaks$start, yy1_meyer_peaks$end))
# creating metaplot DF and running profile_tss
yy1_meyer_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
yy1_meyer_metaplot_df <- profile_tss(yy1_meyer_gr , promoters)
# plotting
ggplot(yy1_meyer_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)

write.csv(yy1_meyer_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//yy1_meyer_peaks_ENCFF328XKC.csv",row.names = TRUE)


#yy1 from Peggy Farnham 
#wget https://www.encodeproject.org/files/ENCFF589PZO/@@download/ENCFF589PZO.bed.gz
yy1_farnham_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/ENCODE_downloads/ENCFF589PZO.bed.gz", col_names = F)

names(yy1_farnham_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')

yy1_farnham_gr <- GRanges(yy1_farnham_peaks$chromosome, IRanges(yy1_farnham_peaks$start, yy1_farnham_peaks$end))
# creating metaplot DF and running profile_tss
yy1_farnham_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
yy1_farnham_metaplot_df <- profile_tss(yy1_farnham_gr , promoters)
# plotting
ggplot(yy1_farnham_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5)

write.csv(yy1_farnham_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//yy1_farnham_peaks_ENCFF589PZO.csv",row.names = TRUE)

#antibody = the two used for the above ChIP experiments have been discontinued by Santa Cruz. They offer this one as a substitute. Sigh. https://www.scbt.com/p/yy1-antibody-h-10?requestFrom=search
```


##Sox2 from Zhou et al. 2016 Cell Res
# doi: 10.1038/cr.2016.15
#Human ES cells and NPCs
#https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4746607/

```{r}
#Sox2 ChIP in Human ESCs 
sox2_hESC_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/Sox2_downloads/GSM1701825_ESC_SOX2.bed.gz", col_names = F)

names(sox2_hESC_peaks) <- c('chromosome', 'start', 'end')

sox2_hESC_gr <- GRanges(sox2_hESC_peaks$chromosome, IRanges(sox2_hESC_peaks$start, sox2_hESC_peaks$end))
# creating metaplot DF and running profile_tss
sox2_hESC_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
sox2_hESC_metaplot_df <- profile_tss(sox2_hESC_gr , promoters)
# plotting
ggplot(sox2_hESC_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5) + ggtitle("Sox2 hESC ChIP Profile Zhou et al. 2016 Cell Res.")

write.csv(sox2_hESC_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//sox2_hESC_peaks.csv",row.names = TRUE)

#Sox2 ChIP in Human NPCs
sox2_hNPC_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/Sox2_downloads/GSM1701828_NPC_SOX2.bed.gz", col_names = F)

names(sox2_hNPC_peaks) <- c('chromosome', 'start', 'end')

sox2_hNPC_gr <- GRanges(sox2_hNPC_peaks$chromosome, IRanges(sox2_hNPC_peaks$start, sox2_hNPC_peaks$end))
# creating metaplot DF and running profile_tss
sox2_hNPC_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
sox2_hNPC_metaplot_df <- profile_tss(sox2_hNPC_gr , promoters)
# plotting
ggplot(sox2_hNPC_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5) + ggtitle("Sox2 hNPC ChIP Profile Zhou et al. 2016 Cell Res.")

write.csv(sox2_hNPC_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//sox2_hNPC_peaks.csv",row.names = TRUE)

##Sox2 in mESCs from Liu et al. 2017 Cell Reports DOI:https://doi.org/10.1016/j.celrep.2017.04.067
#GEO Acquisition Number GSE92846.

sox2_mESC_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/Sox2_downloads/GSM2438473_EC-EA-3702-SOX2_ASYN_1.narrowPeak(1).gz", col_names = F)

names(sox2_mESC_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')

sox2_mESC_gr <- GRanges(sox2_mESC_peaks$chromosome, IRanges(sox2_mESC_peaks$start, sox2_mESC_peaks$end))
# creating metaplot DF and running profile_tss
sox2_mESC_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
sox2_mESC_metaplot_df <- profile_tss(sox2_mESC_gr , promoters)
# plotting
ggplot(sox2_mESC_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5) + ggtitle("Sox2 mESC ChIP Profile Liu et al. 2017 Cell Rep.")

write.csv(sox2_mESC_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//sox2_mESC_Liu2017_peaks.csv",row.names = TRUE)

##Another Sox2 in mESCs 
##https://www.sciencedirect.com/science/article/pii/S0092867416317305?via%3Dihub#fig1
##Chronis et al. 2017 Cell https://doi.org/10.1016/j.cell.2016.12.016
##They also have ATAC Seq 

sox2_mESC_Chronis_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/Sox2_downloads/GSM2417143_ESC_Sox2_Chronis_Cell_2017.bed.gz", col_names = F)

names(sox2_mESC_Chronis_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')

sox2_mESC_Chronis_gr <- GRanges(sox2_mESC_Chronis_peaks$chromosome, IRanges(sox2_mESC_Chronis_peaks$start, sox2_mESC_Chronis_peaks$end))
# creating metaplot DF and running profile_tss
sox2_mESC_Chronis_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
sox2_mESC_Chronis_metaplot_df <- profile_tss(sox2_mESC_Chronis_gr , promoters)
# plotting
ggplot(sox2_mESC_Chronis_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5) + ggtitle("Sox2 mESC ChIP Profile Chronis et al. 2017 Cell")

write.csv(sox2_mESC_Chronis_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//sox2_mESC_Chronis2017_peaks.csv",row.names = TRUE)


##Kim et al. 2018 Nat. Comm. Publication That Desmond really likes
##https://doi.org/10.1038/s41467-018-04818-0
## GEO accession number: GSE113915
sox2_mESC_Kim_peaks <- read_tsv("/Users/teva8543/scratch/rChIP/Sox2_downloads/GSM3123485_Sox2_WT_summits.bed.gz", col_names = F)

names(sox2_mESC_Kim_peaks) <- c('chromosome', 'start', 'end', 'name', 'score', 'strand', 
                       'signalValue', 'pValue', 'qValue', 'peak_center')

sox2_mESC_Kim_gr <- GRanges(sox2_mESC_Kim_peaks$chromosome, IRanges(sox2_mESC_Kim_peaks$start, sox2_mESC_Kim_peaks$end))
# creating metaplot DF and running profile_tss
sox2_mESC_Kim_metaplot_df <- data.frame(x = integer(), dens = numeric(), dbp = character())
sox2_mESC_Kim_metaplot_df <- profile_tss(sox2_mESC_Kim_gr , promoters)
# plotting
ggplot(sox2_mESC_Kim_metaplot_df, aes(x = x, y = dens)) + 
  geom_line(size = 1.5) + ggtitle("Sox2 mESC ChIP Profile Kim et al. 2018 Nat Comm")

write.csv(sox2_mESC_Kim_peaks,"/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_08_08_rChIPCandidateExamination//sox2_mESC_Kim2018_peaks.csv",row.names = TRUE)
```














# goal qPCR peaks to design primers for rchip.
the next steps are:

1) Sort on qValue -log10(FDR) -- thus bigger number more sig
2) Check in browser for peak quality etc... may take a few tries
-copy and paste peak coordinates and then look and see
3) note candidate regions to be tested by qPCR +/- RNase

# The browser is down?? Can't find them -- moving to meta plot