---
title: "2023_02_20_RPB1_rChIP_analysis"
output: html_document
author: Tenaya Vallery
Date: "Feb 20, 2023"
Modified from boulderrinnlab  rchip/analysis/02_RBP1/02_RBP1_analysis.Rmd 
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
#library(ComplexHeatmap)
library(Rsubread)
library(DESeq2)
source("/scratch/Shares/rinn/tenaya/util/intersect_functions.R")
library(ggpubr)
source("/scratch/Shares/rinn/tenaya/util/_setup.R")
source("/scratch/Shares/rinn/tenaya/util/rinnlab_plot_theme.R")

```

# loading in gencode annotations
```{r gencode annotations}
# loading in gencode annotations
gencode_gr <- rtracklayer::import("/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/gencode.v32.annotation.gtf")
# creating mRNA and lncRNA promoters (3kb up and downstream)
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))
# export to save for later
rtracklayer::export(lncrna_mrna_promoters, "/scratch/Shares/rinn/tenaya/rChIP/rChIP/gene_annotations/lncRNA_mrna_promoters.gtf")
```

```{r gencode annotations}
# DEseq2 needs a specific package of 'locfit' that has to be installed first
install.packages("https://cran.r-project.org/src/contrib/Archive/locfit/locfit_1.5-9.4.tar.gz", repos=NULL, type="source")
# Now we can install DEseq2
BiocManager::install("DESeq2")
# now load the package
library(DESeq2)
```


```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("Rsubread")
```


# Importing RBP1 broad_peaks from NF_core pipeline and bam files
```{r RBP1 braod peak files from NF_Core }
# loading in the RBP1 chip without RNAse (-)
RPB1_no_rnase_1 <- rtracklayer::import("/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/macs/broadPeak/RPB1_NoR_R1_peaks.broadPeak")
RPB1_no_ranase_1_df <- RPB1_no_rnase_1 %>% as.data.frame()
# making a fake replicate for deseq
RPB1_no_rnase_2 <- rtracklayer::import("/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/macs/broadPeak/RPB1_NoR_R2_peaks.broadPeak")
# loading in +rnase peak files
RPB1_rnase_1 <- rtracklayer::import("/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/macs/broadPeak/RPB1_Rnase_R1_peaks.broadPeak")
# making a fake replicate for deseq
RPB1_rnase_2 <- rtracklayer::import("/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/macs/broadPeak/RPB1_Rnase_R2_peaks.broadPeak")
# finding fold change between +/- RNase
Rpb1_peaks <- RPB1_no_rnase_1
# creating .saf
RPB1_saf <- Rpb1_peaks %>%
  as.data.frame() %>%
  # making sure to only grab normal chr not contigs (e.g., GL000009.2)
  filter(grepl("chr", seqnames)) %>%
  # renaming cols to fit format of .saf
  dplyr::rename(GeneID = name,
                Chr = seqnames,
                Start = start,
                End = end,
                Strand = strand) %>%
    dplyr::select(GeneID, Chr, Start, End, Strand)
# Writting out .saf
write.table(RPB1_saf, file = "RPB1_peaks_combined.saf", sep = "\t", quote = F, row.names = F)

```

Running feature counts
```{r}
# these files are far away so creating base-path 
base_path <- "/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/"
# Creating a list of all bams and file paths
pe_bams <- c(paste0(base_path, "RPB1_NoR_R1.mLb.clN.sorted.bam"), paste0(base_path, "RPB1_NoR_R2.mLb.clN.sorted.bam"), paste0(base_path, "RPB1_Rnase_R1.mLb.clN.sorted.bam"), paste0(base_path, "RPB1_Rnase_R2.mLb.clN.sorted.bam"))
        
# Running feature counts on CTCF chip samples and inputs (June 2022)
fc_pe <- featureCounts(files = pe_bams,
                    annot.ext = "RPB1_peaks_combined.saf",
                    isPairedEnd = TRUE,
                    nthreads = 6)
# find how many reads in bam file
```

```{r}
counts_df <- fc_pe$counts %>%
  as.data.frame() %>%
  rownames_to_column("peak_id")
  
counts_summary <- counts_df %>%
  pivot_longer(2:3, names_to = "sample_id", values_to = "count") %>%
  group_by(sample_id) %>%
  summarize(total_counts_over_peaks = sum(count))
  
ggplot(counts_df, aes(x = log10(RPB1.NoR.R1.mLb.clN.sorted.bam), 
                      y = log10(RPB1.Rnase.R1.mLb.clN.sorted.bam))) +
  geom_point()

counts_df <- counts_df %>%
  mutate(unnormalized_ratio = log2(RPB1.NoR.R1.mLb.clN.sorted.bam / RPB1.Rnase.R1.mLb.clN.sorted.bam)) %>%
  left_join(RPB1_saf %>% dplyr::rename(peak_id = GeneID))
ggplot(counts_df, aes(x = unnormalized_ratio)) +
  geom_density() +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0)
  
#second replicate
ggplot(counts_df, aes(x = log10(RPB1.NoR.R2.mLb.clN.sorted.bam), 
                      y = log10(RPB1.Rnase.R2.mLb.clN.sorted.bam))) +
  geom_point()
  
counts_df <- counts_df %>%
  mutate(unnormalized_ratio = log2(RPB1.NoR.R2.mLb.clN.sorted.bam / RPB1.Rnase.R2.mLb.clN.sorted.bam)) %>%
  left_join(RPB1_saf %>% dplyr::rename(peak_id = GeneID))
ggplot(counts_df, aes(x = unnormalized_ratio)) +
  geom_density() +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0)
```


```{bash}
#BASH -> Go to TERMINAL
module load samtools
samtools view -c -F 260 /scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/RPB1_NoR_R1.mLb.clN.sorted.bam
# This took a hot minute FYI (~5 minutes). 
# X number of reads aligned = 78184432

samtools view -c -F 260 /scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/RPB1_Rnase_R1.mLb.clN.sorted.bam
# X number of reads aligned=74630534

#Replicate 2

samtools view -c -F 260 /scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/RPB1_NoR_R2.mLb.clN.sorted.bam
#Number of reads aligned = 68630096 

samtools view -c -F 260 /scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/results/bwa/mergedLibrary/RPB1_Rnase_R2.mLb.clN.sorted.bam
#Number of reads aligned = 57802320
```

#Working on generating the Fold Change Log 2 (ChIP/Input) with +Rnase on Y-axis and NoRnase on X-axis
#How cool would it be to combine with CTCF info like Tae Young did for the paper!
```{r}
## create an object of all bam files (***including inputs***)
rpb1_bams <- c(paste0(base_path, "RPB1_NoR_R1.mLb.clN.sorted.bam"), paste0(base_path, "RPB1_NoR_R2.mLb.clN.sorted.bam"), paste0(base_path, "RPB1_Rnase_R1.mLb.clN.sorted.bam"), paste0(base_path, "RPB1_Rnase_R2.mLb.clN.sorted.bam"),paste0(base_path,"input_R1.mLb.clN.sorted.bam"), paste0(base_path,"input_R2.mLb.clN.sorted.bam"))

rpb1_feature_counts <- featureCounts(files=rpb1_bams,
                                     annot.ext = "RPB1_peaks_combined.saf",
                                     isPairedEnd = TRUE,
                                     nthreads = 6)

write_rds(rpb1_feature_counts, "RPB1_feature_counts_output.rds")
```

#new type of r space. 
#Not sure of the benefit for the below. 
```{r start point}
# saving env objects:
# save(chip_rnase_data, counts_matrix, ctcf_feature_counts, ctcf_peaks, ctcf_saf, rnase_input_res, rnase_res, chip_input_res, sample_df, samples_chip, samples_rnase, samples_rnase_input, file = "results/ctcf_chip_combined_5_reps/ctcf_rnase.RData")
# load("results/ctcf_chip_combined_5_reps/ctcf_rnase.RData", verbose = T)
# saving minimal objects 
save(rpb1_bams, RPB1_saf, file = "minimal_env_object.RData")
# load in 
load("minimal_env_object.RData", verbose = T)
```

#Creating sample_df for DEseq input
```{r sample sheet for Deseq}
# setting up samples_df data.frame - start with ctcf_feature_counts object
sample_df <-
  data.frame(filename = fc_pe[["targets"]]) %>%
  mutate(sample_name = filename)
sample_df$sample_name = gsub(".bam|.mLb.clN.sorted.bam|R1.mLb.clN.sorted.bam",
                             "",
                             sample_df$sample_name) 
# mutate to new col "condition"
# creating sample sheet - file names are in target's col
sample_df <- sample_df %>%
  mutate(condition = ifelse(
    grepl("wash", sample_name),
    "rnase_wash",
    ifelse(
      grepl("chip", sample_name),
      "ctcf_chip",
      ifelse(grepl("input", sample_name), "input", "rnase")
    )
  ))
```
