---
title: "RBP1 Chipseq analysis"
author: "JR"
date: "6/29/2022"
output: html_document
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(GenomicRanges)
library(ComplexHeatmap)
library(Rsubread)
library(DESeq2)
source("/scratch/Shares/rinn/tenaya/util/intersect_functions.R")


```

# loading in gencode annotations
```{r gencode annotations}

# loading in gencode annotations
gencode_gr <- rtracklayer::import("/scratch/Shares/rinn/genomes/Homo_sapiens/Gencode/v32/gencode.v32.annotation.gtf")

# creating mRNA and lncRNA promoters (3kb up and downstream)
lncrna_mrna_promoters <- get_promoter_regions(gencode_gr, biotype = c("lncRNA", "protein_coding"))

# export to save for later
rtracklayer::export(lncrna_mrna_promoters, "/scratch/Shares/rinn/JR/rchip/data/gene_annotations/lncRNA_mrna_promoters.gtf")

```

# Importing RBP1 broad_peaks from NF_core pipeline and bam files
```{r RBP1 braod peak files from NF_Core }

# loading in the RBP1 chip without RNAse (-)
RBP1_no_rnase_1 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/RBP1_no_rnase_R1_peaks.broadPeak")

RBP1_no_ranase_1_df <- RBP1_no_rnase_1 %>% as.data.frame()

# making a fake replicate for deseq
RBP1_no_rnase_2 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/RBP1_no_rnase_R1_peaks.broadPeak")

# loading in +rnase peak files
RBP1_rnase_1 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/RBP1_rnase_R1_peaks.broadPeak")

# making a fake replicate for deseq
RBP1_rnase_2 <- rtracklayer::import("/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/macs/broadPeak/RBP1_rnase_R1_peaks.broadPeak")


# finding fold change between +/- RNase

Rbp1_peaks <- RBP1_no_rnase_1

# creating .saf

RBP1_saf <- Rbp1_peaks %>%
  as.data.frame() %>%
  # making sure to only grab normal chr not contigs (e.g., GL000009.2)
  filter(grepl("chr", seqnames)) %>%
  # renaming cols to fit format of .saf
  dplyr::rename(GeneID = name,
                Chr = seqnames,
                Start = start,
                End = end,
                Strand = strand) %>%
    dplyr::select(GeneID, Chr, Start, End, Strand)

# Writting out .saf
write.table(RBP1_saf, "/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/Johns_Results/RBP1_peaks_combined.saf", sep = "\t", quote = F, row.names = F)





```

Running feature counts
```{r}
# these files are far away so creating base-path 
base_path <- "/scratch/Shares/rinn/JR/rchip/data/1-2-2_nextflow_run_all_samples/results/bwa/mergedLibrary/"


# Creating a list of all bams and file paths
pe_bams <- c(paste0(base_path, "RBP1_no_rnase_R1.mLb.clN.sorted.bam"), 
             paste0(base_path, "RBP1_rnase_R1.mLb.clN.sorted.bam"))
        

# Running feature counts on CTCF chip samples and inputs (June 2022)

fc_pe <- featureCounts(files = pe_bams,
                    annot.ext = "/scratch/Shares/rinn/tenaya/rChIP/rChIP/2022_10_04_RPB1/Johns_Results/RBP1_peaks_combined.saf",
                    isPairedEnd = TRUE,
                    nthreads = 6)
# find how many reads in bam file
```

```{r}
counts_df <- fc_pe$counts %>%
  as.data.frame() %>%
  rownames_to_column("peak_id")

counts_summary <- counts_df %>%
  pivot_longer(2:3, names_to = "sample_id", values_to = "count") %>%
  group_by(sample_id) %>%
  summarize(total_counts_over_peaks = sum(count))

ggplot(counts_df, aes(x = log10(RBP1.no.rnase.R1.mLb.clN.sorted.bam), 
                      y = log10(RBP1.rnase.R1.mLb.clN.sorted.bam))) +
  geom_point()




counts_df <- counts_df %>%
  mutate(unnormalized_ratio = log2(RBP1.no.rnase.R1.mLb.clN.sorted.bam / RBP1.rnase.R1.mLb.clN.sorted.bam)) %>%
  left_join(RBP1_saf %>% dplyr::rename(peak_id = GeneID))

ggplot(counts_df, aes(x = unnormalized_ratio)) +
  geom_density() +
  geom_hline(yintercept = 0) +
  geom_vline(xintercept = 0)
```


```{bash}
module load samtools
samtools view -c -F 260 RBP1_no_rnase_R1.mLb.clN.sorted.bam
# X number of reads aligned
91076474

samtools view -c -F 260 RBP1_rnase_R1.mLb.clN.sorted.bam
# X number of reads aligned
87606172




```














